From 9406f6771789794351330a856f93ec694388b419 Mon Sep 17 00:00:00 2001
From: Zhe Wang <zhe.wang@amlogic.com>
Date: Tue, 18 Aug 2015 16:45:18 +0800
Subject: [PATCH 5771/5965] PD#111221: audio: add hw resample lock and unlock
 interface

In G9tv platform, when ddp raw data in hdmi-in,
hw resample must be locked.

Change-Id: I914f6920e78020a7ea9ed80c5867374bc1fa9430
---
 sound/soc/aml/m8/aml_g9tv.c | 18 +++++++++++++++---
 1 file changed, 15 insertions(+), 3 deletions(-)

diff --git a/sound/soc/aml/m8/aml_g9tv.c b/sound/soc/aml/m8/aml_g9tv.c
index d9cebdcf26a4..3c84a9459019 100644
--- a/sound/soc/aml/m8/aml_g9tv.c
+++ b/sound/soc/aml/m8/aml_g9tv.c
@@ -253,6 +253,7 @@ static int aml_spdif_audio_type_set_enum(struct snd_kcontrol *kcontrol,
     return 0;
 }
 
+static int hardware_resample_locked_flag = 0;
 #define RESAMPLE_BUFFER_SOURCE 1 //audioin fifo0
 #define RESAMPLE_CNT_CONTROL 255 //Cnt_ctrl = mclk/fs_out-1 ; fest 256fs
 static int hardware_resample_enable(int input_sr) {
@@ -260,6 +261,10 @@ static int hardware_resample_enable(int input_sr) {
     unsigned long clk_rate = 0;
     unsigned short Avg_cnt_init = 0;
 
+    if (hardware_resample_locked_flag == 1) {
+        return 0;
+    }
+
     if (input_sr < 8000 || input_sr > 48000) {
         printk(KERN_INFO "Error input sample rate, you must set sr first!\n");
         return -1;
@@ -288,7 +293,7 @@ static int hardware_resample_disable(void) {
 }
 
 static const char *hardware_resample_texts[] = { "Disable", "Enable:48K",
-        "Enable:44K", "Enable:32K" };
+        "Enable:44K", "Enable:32K", "Lock Resample", "Unlock Resample"};
 
 static const struct soc_enum hardware_resample_enum = SOC_ENUM_SINGLE(
         SND_SOC_NOPM, 0, ARRAY_SIZE(hardware_resample_texts),
@@ -314,6 +319,14 @@ static int aml_hardware_resample_set_enum(struct snd_kcontrol *kcontrol,
     } else if (ucontrol->value.enumerated.item[0] == 3) {
         hardware_resample_enable(32000);
         aml_audio_Hardware_resample = 3;
+    } else if (ucontrol->value.enumerated.item[0] == 4) {
+        hardware_resample_disable();
+        aml_audio_Hardware_resample = 4;
+        hardware_resample_locked_flag = 1;
+    } else if (ucontrol->value.enumerated.item[0] == 5) {
+        hardware_resample_locked_flag = 0;
+        hardware_resample_enable(48000);
+        aml_audio_Hardware_resample = 5;
     }
     return 0;
 }
@@ -728,8 +741,7 @@ static void aml_g9tv_pinmux_init(struct snd_soc_card *card) {
 
     printk(KERN_INFO "Set audio codec pinmux!\n");
 
-//    if (!IS_MESON_MG9TV_CPU_REVA)
-        p_aml_audio->pin_ctl = devm_pinctrl_get_select(card->dev,
+    p_aml_audio->pin_ctl = devm_pinctrl_get_select(card->dev,
                 "aml_snd_g9tv");
 
     p_audio = p_aml_audio;
-- 
2.19.0

